{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5d6d88c3-0861-4fe1-8388-dfa7b6da0e13",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Attaching SeuratObject\n",
      "\n",
      "\n",
      "Attaching package: ‘Signac’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Seurat’:\n",
      "\n",
      "    FoldChange\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘TFBSTools’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Matrix’:\n",
      "\n",
      "    Matrix\n",
      "\n",
      "\n",
      "Loading required package: BSgenome\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "Loading required package: parallel\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:parallel’:\n",
      "\n",
      "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
      "    clusterExport, clusterMap, parApply, parCapply, parLapply,\n",
      "    parLapplyLB, parRapply, parSapply, parSapplyLB\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:Matrix’:\n",
      "\n",
      "    expand, unname\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: Biostrings\n",
      "\n",
      "Loading required package: XVector\n",
      "\n",
      "\n",
      "Attaching package: ‘Biostrings’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    strsplit\n",
      "\n",
      "\n",
      "Loading required package: rtracklayer\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(Seurat)\n",
    "library(Signac)\n",
    "library(chromVAR)\n",
    "library(motifmatchr)\n",
    "library(Matrix)\n",
    "library(TFBSTools)\n",
    "library(JASPAR2020)\n",
    "library(BSgenome.Rnorvegicus.UCSC.rn6)\n",
    "library(BiocParallel)\n",
    "library(ggplot2)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "370ee590-576f-48bd-86be-f84304f847d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# snATAC data takes about 5-10 mins~ to read  from RDS\n",
    "snatac_rds <- \"harmony_with_predicted_id.rds\"\n",
    "rat_data <- readRDS(snatac_rds)\n",
    "\n",
    "DefaultAssay(rat_data) <- \"peaks\"\n",
    "Idents(rat_data) <- rat_data$predicted.id\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "921afa28-6bb3-4a2e-8e73-8b1a909bd5ef",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Building motif matrix\n",
      "\n",
      "Finding motif positions\n",
      "\n",
      "Creating Motif object\n",
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# Takes 5 mins~\n",
    "# LoadVertebrate Matrix\n",
    "pfm <- getMatrixSet(x = JASPAR2020,\n",
    "                   opts = list(collection = 'CORE', #collection to find\n",
    "                   tax_group = 'vertebrates', #taxonomy group\n",
    "                   all_versions = FALSE)\n",
    "                   )\n",
    "\n",
    "\n",
    "# Add Motif data\n",
    "rat_data <- AddMotifs(\n",
    "    object = rat_data,\n",
    "    genome = BSgenome.Rnorvegicus.UCSC.rn6,\n",
    "    pfm = pfm\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "f8d42a3c-a6a3-48af-b2d3-108c167e24ad",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Computing GC bias per region\n",
      "\n",
      "Selecting background regions\n",
      "\n",
      "Computing deviations from background\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  |======================================================================| 100%\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Constructing chromVAR assay\n",
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# RUN CHROMVAR TAKES A WHILE 1:30 - 2 hours\n",
    "register(MulticoreParam(48, progressbar = TRUE))\n",
    "chromvar <- RunChromVAR(object = rat_data, genome = BSgenome.Rnorvegicus.UCSC.rn6)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "6a8728d1-5032-4193-b882-8fa2f582ed09",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "An object of class Seurat \n",
       "314686 features across 81912 samples within 3 assays \n",
       "Active assay: peaks (291844 features, 290644 variable features)\n",
       " 2 other assays present: RNA, chromvar\n",
       " 3 dimensional reductions calculated: lsi, umap, harmony"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "\n",
    "chromvar # Confirm chromvar assay present\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "id": "24d62cf3-db9c-4dd2-ace2-fb1eb08d3776",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"wilcox\"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 746 × 6</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>p_val</th><th scope=col>avg_diff</th><th scope=col>pct.1</th><th scope=col>pct.2</th><th scope=col>p_val_adj</th><th scope=col>fdr</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>MA0113.3</th><td>2.080285e-26</td><td>-0.3816178</td><td>0.773</td><td>0.832</td><td>1.551892e-23</td><td>1.551892e-23</td></tr>\n",
       "\t<tr><th scope=row>MA0727.1</th><td>5.342531e-26</td><td>-0.3842041</td><td>0.784</td><td>0.837</td><td>3.985528e-23</td><td>1.992764e-23</td></tr>\n",
       "\t<tr><th scope=row>MA0007.3</th><td>6.822374e-24</td><td>-0.3327425</td><td>0.665</td><td>0.747</td><td>5.089491e-21</td><td>1.696497e-21</td></tr>\n",
       "\t<tr><th scope=row>MA0480.1</th><td>1.178422e-19</td><td> 0.2447419</td><td>0.844</td><td>0.765</td><td>8.791027e-17</td><td>2.197757e-17</td></tr>\n",
       "\t<tr><th scope=row>MA1150.1</th><td>5.922017e-17</td><td> 0.3202596</td><td>0.995</td><td>0.993</td><td>4.417824e-14</td><td>8.835649e-15</td></tr>\n",
       "\t<tr><th scope=row>MA0071.1</th><td>1.609840e-15</td><td> 0.2488727</td><td>0.974</td><td>0.957</td><td>1.200940e-12</td><td>2.001567e-13</td></tr>\n",
       "\t<tr><th scope=row>MA1151.1</th><td>1.151912e-13</td><td> 0.2903658</td><td>0.996</td><td>0.993</td><td>8.593261e-11</td><td>1.227609e-11</td></tr>\n",
       "\t<tr><th scope=row>MA0746.2</th><td>8.953663e-13</td><td>-0.1340090</td><td>0.116</td><td>0.185</td><td>6.679433e-10</td><td>8.349291e-11</td></tr>\n",
       "\t<tr><th scope=row>MA0442.2</th><td>1.526744e-12</td><td> 0.1942299</td><td>0.910</td><td>0.859</td><td>1.138951e-09</td><td>1.265501e-10</td></tr>\n",
       "\t<tr><th scope=row>MA0079.4</th><td>3.148802e-12</td><td>-0.1325268</td><td>0.213</td><td>0.284</td><td>2.349006e-09</td><td>2.349006e-10</td></tr>\n",
       "\t<tr><th scope=row>MA1517.1</th><td>6.178909e-12</td><td>-0.1544406</td><td>0.064</td><td>0.123</td><td>4.609466e-09</td><td>4.190424e-10</td></tr>\n",
       "\t<tr><th scope=row>MA0139.1</th><td>9.610053e-12</td><td>-0.1929606</td><td>0.607</td><td>0.691</td><td>7.169099e-09</td><td>5.974249e-10</td></tr>\n",
       "\t<tr><th scope=row>MA0740.1</th><td>1.212124e-11</td><td>-0.1424404</td><td>0.178</td><td>0.248</td><td>9.042443e-09</td><td>6.955725e-10</td></tr>\n",
       "\t<tr><th scope=row>MA0502.2</th><td>2.075256e-11</td><td> 0.1943036</td><td>0.916</td><td>0.872</td><td>1.548141e-08</td><td>1.105815e-09</td></tr>\n",
       "\t<tr><th scope=row>MA0161.2</th><td>2.995669e-11</td><td> 0.2212196</td><td>0.998</td><td>0.993</td><td>2.234769e-08</td><td>1.489846e-09</td></tr>\n",
       "\t<tr><th scope=row>MA1130.1</th><td>4.090674e-11</td><td>-0.2416664</td><td>0.005</td><td>0.020</td><td>3.051643e-08</td><td>1.840580e-09</td></tr>\n",
       "\t<tr><th scope=row>MA0162.4</th><td>4.194350e-11</td><td>-0.1424868</td><td>0.010</td><td>0.020</td><td>3.128985e-08</td><td>1.840580e-09</td></tr>\n",
       "\t<tr><th scope=row>MA1644.1</th><td>5.156797e-11</td><td> 0.1658892</td><td>0.662</td><td>0.590</td><td>3.846971e-08</td><td>2.137206e-09</td></tr>\n",
       "\t<tr><th scope=row>MA0742.1</th><td>6.413254e-11</td><td>-0.1591836</td><td>0.131</td><td>0.199</td><td>4.784288e-08</td><td>2.518046e-09</td></tr>\n",
       "\t<tr><th scope=row>MA0060.3</th><td>1.348471e-10</td><td> 0.1735255</td><td>0.651</td><td>0.587</td><td>1.005960e-07</td><td>4.814636e-09</td></tr>\n",
       "\t<tr><th scope=row>MA0593.1</th><td>1.355326e-10</td><td> 0.1840209</td><td>0.887</td><td>0.833</td><td>1.011074e-07</td><td>4.814636e-09</td></tr>\n",
       "\t<tr><th scope=row>MA0477.2</th><td>1.965481e-10</td><td>-0.2164108</td><td>0.006</td><td>0.020</td><td>1.466249e-07</td><td>6.664766e-09</td></tr>\n",
       "\t<tr><th scope=row>MA1134.1</th><td>2.704817e-10</td><td>-0.2308074</td><td>0.007</td><td>0.020</td><td>2.017793e-07</td><td>8.773014e-09</td></tr>\n",
       "\t<tr><th scope=row>MA0072.1</th><td>2.961261e-10</td><td> 0.2192463</td><td>0.988</td><td>0.979</td><td>2.209101e-07</td><td>9.204587e-09</td></tr>\n",
       "\t<tr><th scope=row>MA1116.1</th><td>3.534273e-10</td><td> 0.1598859</td><td>0.929</td><td>0.886</td><td>2.636568e-07</td><td>1.054627e-08</td></tr>\n",
       "\t<tr><th scope=row>MA1137.1</th><td>3.878982e-10</td><td>-0.2284086</td><td>0.008</td><td>0.021</td><td>2.893721e-07</td><td>1.112970e-08</td></tr>\n",
       "\t<tr><th scope=row>MA0732.1</th><td>5.322822e-10</td><td>-0.1405313</td><td>0.022</td><td>0.038</td><td>3.970825e-07</td><td>1.470676e-08</td></tr>\n",
       "\t<tr><th scope=row>MA0666.1</th><td>8.867248e-10</td><td> 0.1901239</td><td>0.964</td><td>0.952</td><td>6.614967e-07</td><td>2.100933e-08</td></tr>\n",
       "\t<tr><th scope=row>MA0708.1</th><td>8.867248e-10</td><td> 0.1901239</td><td>0.964</td><td>0.952</td><td>6.614967e-07</td><td>2.100933e-08</td></tr>\n",
       "\t<tr><th scope=row>MA0709.1</th><td>8.867248e-10</td><td> 0.1901239</td><td>0.964</td><td>0.952</td><td>6.614967e-07</td><td>2.100933e-08</td></tr>\n",
       "\t<tr><th scope=row>⋮</th><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><th scope=row>MA0814.2</th><td>0.8775416</td><td>-0.0012909307</td><td>0.648</td><td>0.646</td><td>1</td><td>0.9130349</td></tr>\n",
       "\t<tr><th scope=row>MA1607.1</th><td>0.8907774</td><td> 0.0120231042</td><td>0.597</td><td>0.606</td><td>1</td><td>0.9255152</td></tr>\n",
       "\t<tr><th scope=row>MA0626.1</th><td>0.9003377</td><td> 0.0121991967</td><td>0.126</td><td>0.139</td><td>1</td><td>0.9326034</td></tr>\n",
       "\t<tr><th scope=row>MA0781.1</th><td>0.9005587</td><td> 0.0024475142</td><td>0.354</td><td>0.364</td><td>1</td><td>0.9326034</td></tr>\n",
       "\t<tr><th scope=row>MA0879.1</th><td>0.9015468</td><td>-0.0124056172</td><td>0.636</td><td>0.629</td><td>1</td><td>0.9326034</td></tr>\n",
       "\t<tr><th scope=row>MA1569.1</th><td>0.9026001</td><td>-0.0002614462</td><td>0.476</td><td>0.467</td><td>1</td><td>0.9326034</td></tr>\n",
       "\t<tr><th scope=row>MA0689.1</th><td>0.9083895</td><td> 0.0035505919</td><td>0.164</td><td>0.195</td><td>1</td><td>0.9372871</td></tr>\n",
       "\t<tr><th scope=row>MA0069.1</th><td>0.9212343</td><td>-0.0011142453</td><td>0.364</td><td>0.355</td><td>1</td><td>0.9492276</td></tr>\n",
       "\t<tr><th scope=row>MA0819.1</th><td>0.9228512</td><td> 0.0178486823</td><td>0.121</td><td>0.126</td><td>1</td><td>0.9495821</td></tr>\n",
       "\t<tr><th scope=row>MA0815.1</th><td>0.9241424</td><td> 0.0115731168</td><td>0.703</td><td>0.698</td><td>1</td><td>0.9496008</td></tr>\n",
       "\t<tr><th scope=row>MA0839.1</th><td>0.9371164</td><td> 0.0073612132</td><td>0.268</td><td>0.268</td><td>1</td><td>0.9616078</td></tr>\n",
       "\t<tr><th scope=row>MA0024.3</th><td>0.9425644</td><td>-0.0013498868</td><td>0.526</td><td>0.526</td><td>1</td><td>0.9658696</td></tr>\n",
       "\t<tr><th scope=row>MA0003.4</th><td>0.9444987</td><td>-0.0001510493</td><td>0.555</td><td>0.546</td><td>1</td><td>0.9665240</td></tr>\n",
       "\t<tr><th scope=row>MA0811.1</th><td>0.9504210</td><td> 0.0103099361</td><td>0.618</td><td>0.630</td><td>1</td><td>0.9712521</td></tr>\n",
       "\t<tr><th scope=row>MA0037.3</th><td>0.9556658</td><td>-0.0023560906</td><td>0.138</td><td>0.160</td><td>1</td><td>0.9751980</td></tr>\n",
       "\t<tr><th scope=row>MA1419.1</th><td>0.9569086</td><td> 0.0135376607</td><td>0.616</td><td>0.612</td><td>1</td><td>0.9751980</td></tr>\n",
       "\t<tr><th scope=row>MA0102.4</th><td>0.9582039</td><td>-0.0037102403</td><td>0.442</td><td>0.437</td><td>1</td><td>0.9751980</td></tr>\n",
       "\t<tr><th scope=row>MA0767.1</th><td>0.9676795</td><td> 0.0016362936</td><td>0.222</td><td>0.232</td><td>1</td><td>0.9834999</td></tr>\n",
       "\t<tr><th scope=row>MA0028.2</th><td>0.9700230</td><td> 0.0066278163</td><td>0.393</td><td>0.400</td><td>1</td><td>0.9845403</td></tr>\n",
       "\t<tr><th scope=row>MA0749.1</th><td>0.9733881</td><td>-0.0009654235</td><td>0.406</td><td>0.402</td><td>1</td><td>0.9866135</td></tr>\n",
       "\t<tr><th scope=row>MA1563.1</th><td>0.9793600</td><td> 0.0039495215</td><td>0.529</td><td>0.533</td><td>1</td><td>0.9913196</td></tr>\n",
       "\t<tr><th scope=row>MA1552.1</th><td>0.9824250</td><td>-0.0028840138</td><td>0.489</td><td>0.478</td><td>1</td><td>0.9927754</td></tr>\n",
       "\t<tr><th scope=row>MA1471.1</th><td>0.9834598</td><td> 0.0007837921</td><td>0.457</td><td>0.455</td><td>1</td><td>0.9927754</td></tr>\n",
       "\t<tr><th scope=row>MA1483.1</th><td>0.9848484</td><td>-0.0015546475</td><td>0.419</td><td>0.428</td><td>1</td><td>0.9928336</td></tr>\n",
       "\t<tr><th scope=row>MA1114.1</th><td>0.9874160</td><td> 0.0044616199</td><td>0.228</td><td>0.243</td><td>1</td><td>0.9940787</td></tr>\n",
       "\t<tr><th scope=row>MA0631.1</th><td>0.9895514</td><td>-0.0015982223</td><td>0.313</td><td>0.315</td><td>1</td><td>0.9948859</td></tr>\n",
       "\t<tr><th scope=row>MA0847.2</th><td>0.9930888</td><td> 0.0072745558</td><td>0.603</td><td>0.599</td><td>1</td><td>0.9959829</td></tr>\n",
       "\t<tr><th scope=row>MA1576.1</th><td>0.9940583</td><td> 0.0037482974</td><td>0.638</td><td>0.632</td><td>1</td><td>0.9959829</td></tr>\n",
       "\t<tr><th scope=row>MA1554.1</th><td>0.9946478</td><td> 0.0237875198</td><td>0.081</td><td>0.100</td><td>1</td><td>0.9959829</td></tr>\n",
       "\t<tr><th scope=row>MA1573.1</th><td>0.9997445</td><td>-0.0040914399</td><td>0.624</td><td>0.615</td><td>1</td><td>0.9997445</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 746 × 6\n",
       "\\begin{tabular}{r|llllll}\n",
       "  & p\\_val & avg\\_diff & pct.1 & pct.2 & p\\_val\\_adj & fdr\\\\\n",
       "  & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\tMA0113.3 & 2.080285e-26 & -0.3816178 & 0.773 & 0.832 & 1.551892e-23 & 1.551892e-23\\\\\n",
       "\tMA0727.1 & 5.342531e-26 & -0.3842041 & 0.784 & 0.837 & 3.985528e-23 & 1.992764e-23\\\\\n",
       "\tMA0007.3 & 6.822374e-24 & -0.3327425 & 0.665 & 0.747 & 5.089491e-21 & 1.696497e-21\\\\\n",
       "\tMA0480.1 & 1.178422e-19 &  0.2447419 & 0.844 & 0.765 & 8.791027e-17 & 2.197757e-17\\\\\n",
       "\tMA1150.1 & 5.922017e-17 &  0.3202596 & 0.995 & 0.993 & 4.417824e-14 & 8.835649e-15\\\\\n",
       "\tMA0071.1 & 1.609840e-15 &  0.2488727 & 0.974 & 0.957 & 1.200940e-12 & 2.001567e-13\\\\\n",
       "\tMA1151.1 & 1.151912e-13 &  0.2903658 & 0.996 & 0.993 & 8.593261e-11 & 1.227609e-11\\\\\n",
       "\tMA0746.2 & 8.953663e-13 & -0.1340090 & 0.116 & 0.185 & 6.679433e-10 & 8.349291e-11\\\\\n",
       "\tMA0442.2 & 1.526744e-12 &  0.1942299 & 0.910 & 0.859 & 1.138951e-09 & 1.265501e-10\\\\\n",
       "\tMA0079.4 & 3.148802e-12 & -0.1325268 & 0.213 & 0.284 & 2.349006e-09 & 2.349006e-10\\\\\n",
       "\tMA1517.1 & 6.178909e-12 & -0.1544406 & 0.064 & 0.123 & 4.609466e-09 & 4.190424e-10\\\\\n",
       "\tMA0139.1 & 9.610053e-12 & -0.1929606 & 0.607 & 0.691 & 7.169099e-09 & 5.974249e-10\\\\\n",
       "\tMA0740.1 & 1.212124e-11 & -0.1424404 & 0.178 & 0.248 & 9.042443e-09 & 6.955725e-10\\\\\n",
       "\tMA0502.2 & 2.075256e-11 &  0.1943036 & 0.916 & 0.872 & 1.548141e-08 & 1.105815e-09\\\\\n",
       "\tMA0161.2 & 2.995669e-11 &  0.2212196 & 0.998 & 0.993 & 2.234769e-08 & 1.489846e-09\\\\\n",
       "\tMA1130.1 & 4.090674e-11 & -0.2416664 & 0.005 & 0.020 & 3.051643e-08 & 1.840580e-09\\\\\n",
       "\tMA0162.4 & 4.194350e-11 & -0.1424868 & 0.010 & 0.020 & 3.128985e-08 & 1.840580e-09\\\\\n",
       "\tMA1644.1 & 5.156797e-11 &  0.1658892 & 0.662 & 0.590 & 3.846971e-08 & 2.137206e-09\\\\\n",
       "\tMA0742.1 & 6.413254e-11 & -0.1591836 & 0.131 & 0.199 & 4.784288e-08 & 2.518046e-09\\\\\n",
       "\tMA0060.3 & 1.348471e-10 &  0.1735255 & 0.651 & 0.587 & 1.005960e-07 & 4.814636e-09\\\\\n",
       "\tMA0593.1 & 1.355326e-10 &  0.1840209 & 0.887 & 0.833 & 1.011074e-07 & 4.814636e-09\\\\\n",
       "\tMA0477.2 & 1.965481e-10 & -0.2164108 & 0.006 & 0.020 & 1.466249e-07 & 6.664766e-09\\\\\n",
       "\tMA1134.1 & 2.704817e-10 & -0.2308074 & 0.007 & 0.020 & 2.017793e-07 & 8.773014e-09\\\\\n",
       "\tMA0072.1 & 2.961261e-10 &  0.2192463 & 0.988 & 0.979 & 2.209101e-07 & 9.204587e-09\\\\\n",
       "\tMA1116.1 & 3.534273e-10 &  0.1598859 & 0.929 & 0.886 & 2.636568e-07 & 1.054627e-08\\\\\n",
       "\tMA1137.1 & 3.878982e-10 & -0.2284086 & 0.008 & 0.021 & 2.893721e-07 & 1.112970e-08\\\\\n",
       "\tMA0732.1 & 5.322822e-10 & -0.1405313 & 0.022 & 0.038 & 3.970825e-07 & 1.470676e-08\\\\\n",
       "\tMA0666.1 & 8.867248e-10 &  0.1901239 & 0.964 & 0.952 & 6.614967e-07 & 2.100933e-08\\\\\n",
       "\tMA0708.1 & 8.867248e-10 &  0.1901239 & 0.964 & 0.952 & 6.614967e-07 & 2.100933e-08\\\\\n",
       "\tMA0709.1 & 8.867248e-10 &  0.1901239 & 0.964 & 0.952 & 6.614967e-07 & 2.100933e-08\\\\\n",
       "\t⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\tMA0814.2 & 0.8775416 & -0.0012909307 & 0.648 & 0.646 & 1 & 0.9130349\\\\\n",
       "\tMA1607.1 & 0.8907774 &  0.0120231042 & 0.597 & 0.606 & 1 & 0.9255152\\\\\n",
       "\tMA0626.1 & 0.9003377 &  0.0121991967 & 0.126 & 0.139 & 1 & 0.9326034\\\\\n",
       "\tMA0781.1 & 0.9005587 &  0.0024475142 & 0.354 & 0.364 & 1 & 0.9326034\\\\\n",
       "\tMA0879.1 & 0.9015468 & -0.0124056172 & 0.636 & 0.629 & 1 & 0.9326034\\\\\n",
       "\tMA1569.1 & 0.9026001 & -0.0002614462 & 0.476 & 0.467 & 1 & 0.9326034\\\\\n",
       "\tMA0689.1 & 0.9083895 &  0.0035505919 & 0.164 & 0.195 & 1 & 0.9372871\\\\\n",
       "\tMA0069.1 & 0.9212343 & -0.0011142453 & 0.364 & 0.355 & 1 & 0.9492276\\\\\n",
       "\tMA0819.1 & 0.9228512 &  0.0178486823 & 0.121 & 0.126 & 1 & 0.9495821\\\\\n",
       "\tMA0815.1 & 0.9241424 &  0.0115731168 & 0.703 & 0.698 & 1 & 0.9496008\\\\\n",
       "\tMA0839.1 & 0.9371164 &  0.0073612132 & 0.268 & 0.268 & 1 & 0.9616078\\\\\n",
       "\tMA0024.3 & 0.9425644 & -0.0013498868 & 0.526 & 0.526 & 1 & 0.9658696\\\\\n",
       "\tMA0003.4 & 0.9444987 & -0.0001510493 & 0.555 & 0.546 & 1 & 0.9665240\\\\\n",
       "\tMA0811.1 & 0.9504210 &  0.0103099361 & 0.618 & 0.630 & 1 & 0.9712521\\\\\n",
       "\tMA0037.3 & 0.9556658 & -0.0023560906 & 0.138 & 0.160 & 1 & 0.9751980\\\\\n",
       "\tMA1419.1 & 0.9569086 &  0.0135376607 & 0.616 & 0.612 & 1 & 0.9751980\\\\\n",
       "\tMA0102.4 & 0.9582039 & -0.0037102403 & 0.442 & 0.437 & 1 & 0.9751980\\\\\n",
       "\tMA0767.1 & 0.9676795 &  0.0016362936 & 0.222 & 0.232 & 1 & 0.9834999\\\\\n",
       "\tMA0028.2 & 0.9700230 &  0.0066278163 & 0.393 & 0.400 & 1 & 0.9845403\\\\\n",
       "\tMA0749.1 & 0.9733881 & -0.0009654235 & 0.406 & 0.402 & 1 & 0.9866135\\\\\n",
       "\tMA1563.1 & 0.9793600 &  0.0039495215 & 0.529 & 0.533 & 1 & 0.9913196\\\\\n",
       "\tMA1552.1 & 0.9824250 & -0.0028840138 & 0.489 & 0.478 & 1 & 0.9927754\\\\\n",
       "\tMA1471.1 & 0.9834598 &  0.0007837921 & 0.457 & 0.455 & 1 & 0.9927754\\\\\n",
       "\tMA1483.1 & 0.9848484 & -0.0015546475 & 0.419 & 0.428 & 1 & 0.9928336\\\\\n",
       "\tMA1114.1 & 0.9874160 &  0.0044616199 & 0.228 & 0.243 & 1 & 0.9940787\\\\\n",
       "\tMA0631.1 & 0.9895514 & -0.0015982223 & 0.313 & 0.315 & 1 & 0.9948859\\\\\n",
       "\tMA0847.2 & 0.9930888 &  0.0072745558 & 0.603 & 0.599 & 1 & 0.9959829\\\\\n",
       "\tMA1576.1 & 0.9940583 &  0.0037482974 & 0.638 & 0.632 & 1 & 0.9959829\\\\\n",
       "\tMA1554.1 & 0.9946478 &  0.0237875198 & 0.081 & 0.100 & 1 & 0.9959829\\\\\n",
       "\tMA1573.1 & 0.9997445 & -0.0040914399 & 0.624 & 0.615 & 1 & 0.9997445\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 746 × 6\n",
       "\n",
       "| <!--/--> | p_val &lt;dbl&gt; | avg_diff &lt;dbl&gt; | pct.1 &lt;dbl&gt; | pct.2 &lt;dbl&gt; | p_val_adj &lt;dbl&gt; | fdr &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|\n",
       "| MA0113.3 | 2.080285e-26 | -0.3816178 | 0.773 | 0.832 | 1.551892e-23 | 1.551892e-23 |\n",
       "| MA0727.1 | 5.342531e-26 | -0.3842041 | 0.784 | 0.837 | 3.985528e-23 | 1.992764e-23 |\n",
       "| MA0007.3 | 6.822374e-24 | -0.3327425 | 0.665 | 0.747 | 5.089491e-21 | 1.696497e-21 |\n",
       "| MA0480.1 | 1.178422e-19 |  0.2447419 | 0.844 | 0.765 | 8.791027e-17 | 2.197757e-17 |\n",
       "| MA1150.1 | 5.922017e-17 |  0.3202596 | 0.995 | 0.993 | 4.417824e-14 | 8.835649e-15 |\n",
       "| MA0071.1 | 1.609840e-15 |  0.2488727 | 0.974 | 0.957 | 1.200940e-12 | 2.001567e-13 |\n",
       "| MA1151.1 | 1.151912e-13 |  0.2903658 | 0.996 | 0.993 | 8.593261e-11 | 1.227609e-11 |\n",
       "| MA0746.2 | 8.953663e-13 | -0.1340090 | 0.116 | 0.185 | 6.679433e-10 | 8.349291e-11 |\n",
       "| MA0442.2 | 1.526744e-12 |  0.1942299 | 0.910 | 0.859 | 1.138951e-09 | 1.265501e-10 |\n",
       "| MA0079.4 | 3.148802e-12 | -0.1325268 | 0.213 | 0.284 | 2.349006e-09 | 2.349006e-10 |\n",
       "| MA1517.1 | 6.178909e-12 | -0.1544406 | 0.064 | 0.123 | 4.609466e-09 | 4.190424e-10 |\n",
       "| MA0139.1 | 9.610053e-12 | -0.1929606 | 0.607 | 0.691 | 7.169099e-09 | 5.974249e-10 |\n",
       "| MA0740.1 | 1.212124e-11 | -0.1424404 | 0.178 | 0.248 | 9.042443e-09 | 6.955725e-10 |\n",
       "| MA0502.2 | 2.075256e-11 |  0.1943036 | 0.916 | 0.872 | 1.548141e-08 | 1.105815e-09 |\n",
       "| MA0161.2 | 2.995669e-11 |  0.2212196 | 0.998 | 0.993 | 2.234769e-08 | 1.489846e-09 |\n",
       "| MA1130.1 | 4.090674e-11 | -0.2416664 | 0.005 | 0.020 | 3.051643e-08 | 1.840580e-09 |\n",
       "| MA0162.4 | 4.194350e-11 | -0.1424868 | 0.010 | 0.020 | 3.128985e-08 | 1.840580e-09 |\n",
       "| MA1644.1 | 5.156797e-11 |  0.1658892 | 0.662 | 0.590 | 3.846971e-08 | 2.137206e-09 |\n",
       "| MA0742.1 | 6.413254e-11 | -0.1591836 | 0.131 | 0.199 | 4.784288e-08 | 2.518046e-09 |\n",
       "| MA0060.3 | 1.348471e-10 |  0.1735255 | 0.651 | 0.587 | 1.005960e-07 | 4.814636e-09 |\n",
       "| MA0593.1 | 1.355326e-10 |  0.1840209 | 0.887 | 0.833 | 1.011074e-07 | 4.814636e-09 |\n",
       "| MA0477.2 | 1.965481e-10 | -0.2164108 | 0.006 | 0.020 | 1.466249e-07 | 6.664766e-09 |\n",
       "| MA1134.1 | 2.704817e-10 | -0.2308074 | 0.007 | 0.020 | 2.017793e-07 | 8.773014e-09 |\n",
       "| MA0072.1 | 2.961261e-10 |  0.2192463 | 0.988 | 0.979 | 2.209101e-07 | 9.204587e-09 |\n",
       "| MA1116.1 | 3.534273e-10 |  0.1598859 | 0.929 | 0.886 | 2.636568e-07 | 1.054627e-08 |\n",
       "| MA1137.1 | 3.878982e-10 | -0.2284086 | 0.008 | 0.021 | 2.893721e-07 | 1.112970e-08 |\n",
       "| MA0732.1 | 5.322822e-10 | -0.1405313 | 0.022 | 0.038 | 3.970825e-07 | 1.470676e-08 |\n",
       "| MA0666.1 | 8.867248e-10 |  0.1901239 | 0.964 | 0.952 | 6.614967e-07 | 2.100933e-08 |\n",
       "| MA0708.1 | 8.867248e-10 |  0.1901239 | 0.964 | 0.952 | 6.614967e-07 | 2.100933e-08 |\n",
       "| MA0709.1 | 8.867248e-10 |  0.1901239 | 0.964 | 0.952 | 6.614967e-07 | 2.100933e-08 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| MA0814.2 | 0.8775416 | -0.0012909307 | 0.648 | 0.646 | 1 | 0.9130349 |\n",
       "| MA1607.1 | 0.8907774 |  0.0120231042 | 0.597 | 0.606 | 1 | 0.9255152 |\n",
       "| MA0626.1 | 0.9003377 |  0.0121991967 | 0.126 | 0.139 | 1 | 0.9326034 |\n",
       "| MA0781.1 | 0.9005587 |  0.0024475142 | 0.354 | 0.364 | 1 | 0.9326034 |\n",
       "| MA0879.1 | 0.9015468 | -0.0124056172 | 0.636 | 0.629 | 1 | 0.9326034 |\n",
       "| MA1569.1 | 0.9026001 | -0.0002614462 | 0.476 | 0.467 | 1 | 0.9326034 |\n",
       "| MA0689.1 | 0.9083895 |  0.0035505919 | 0.164 | 0.195 | 1 | 0.9372871 |\n",
       "| MA0069.1 | 0.9212343 | -0.0011142453 | 0.364 | 0.355 | 1 | 0.9492276 |\n",
       "| MA0819.1 | 0.9228512 |  0.0178486823 | 0.121 | 0.126 | 1 | 0.9495821 |\n",
       "| MA0815.1 | 0.9241424 |  0.0115731168 | 0.703 | 0.698 | 1 | 0.9496008 |\n",
       "| MA0839.1 | 0.9371164 |  0.0073612132 | 0.268 | 0.268 | 1 | 0.9616078 |\n",
       "| MA0024.3 | 0.9425644 | -0.0013498868 | 0.526 | 0.526 | 1 | 0.9658696 |\n",
       "| MA0003.4 | 0.9444987 | -0.0001510493 | 0.555 | 0.546 | 1 | 0.9665240 |\n",
       "| MA0811.1 | 0.9504210 |  0.0103099361 | 0.618 | 0.630 | 1 | 0.9712521 |\n",
       "| MA0037.3 | 0.9556658 | -0.0023560906 | 0.138 | 0.160 | 1 | 0.9751980 |\n",
       "| MA1419.1 | 0.9569086 |  0.0135376607 | 0.616 | 0.612 | 1 | 0.9751980 |\n",
       "| MA0102.4 | 0.9582039 | -0.0037102403 | 0.442 | 0.437 | 1 | 0.9751980 |\n",
       "| MA0767.1 | 0.9676795 |  0.0016362936 | 0.222 | 0.232 | 1 | 0.9834999 |\n",
       "| MA0028.2 | 0.9700230 |  0.0066278163 | 0.393 | 0.400 | 1 | 0.9845403 |\n",
       "| MA0749.1 | 0.9733881 | -0.0009654235 | 0.406 | 0.402 | 1 | 0.9866135 |\n",
       "| MA1563.1 | 0.9793600 |  0.0039495215 | 0.529 | 0.533 | 1 | 0.9913196 |\n",
       "| MA1552.1 | 0.9824250 | -0.0028840138 | 0.489 | 0.478 | 1 | 0.9927754 |\n",
       "| MA1471.1 | 0.9834598 |  0.0007837921 | 0.457 | 0.455 | 1 | 0.9927754 |\n",
       "| MA1483.1 | 0.9848484 | -0.0015546475 | 0.419 | 0.428 | 1 | 0.9928336 |\n",
       "| MA1114.1 | 0.9874160 |  0.0044616199 | 0.228 | 0.243 | 1 | 0.9940787 |\n",
       "| MA0631.1 | 0.9895514 | -0.0015982223 | 0.313 | 0.315 | 1 | 0.9948859 |\n",
       "| MA0847.2 | 0.9930888 |  0.0072745558 | 0.603 | 0.599 | 1 | 0.9959829 |\n",
       "| MA1576.1 | 0.9940583 |  0.0037482974 | 0.638 | 0.632 | 1 | 0.9959829 |\n",
       "| MA1554.1 | 0.9946478 |  0.0237875198 | 0.081 | 0.100 | 1 | 0.9959829 |\n",
       "| MA1573.1 | 0.9997445 | -0.0040914399 | 0.624 | 0.615 | 1 | 0.9997445 |\n",
       "\n"
      ],
      "text/plain": [
       "         p_val        avg_diff      pct.1 pct.2 p_val_adj    fdr         \n",
       "MA0113.3 2.080285e-26 -0.3816178    0.773 0.832 1.551892e-23 1.551892e-23\n",
       "MA0727.1 5.342531e-26 -0.3842041    0.784 0.837 3.985528e-23 1.992764e-23\n",
       "MA0007.3 6.822374e-24 -0.3327425    0.665 0.747 5.089491e-21 1.696497e-21\n",
       "MA0480.1 1.178422e-19  0.2447419    0.844 0.765 8.791027e-17 2.197757e-17\n",
       "MA1150.1 5.922017e-17  0.3202596    0.995 0.993 4.417824e-14 8.835649e-15\n",
       "MA0071.1 1.609840e-15  0.2488727    0.974 0.957 1.200940e-12 2.001567e-13\n",
       "MA1151.1 1.151912e-13  0.2903658    0.996 0.993 8.593261e-11 1.227609e-11\n",
       "MA0746.2 8.953663e-13 -0.1340090    0.116 0.185 6.679433e-10 8.349291e-11\n",
       "MA0442.2 1.526744e-12  0.1942299    0.910 0.859 1.138951e-09 1.265501e-10\n",
       "MA0079.4 3.148802e-12 -0.1325268    0.213 0.284 2.349006e-09 2.349006e-10\n",
       "MA1517.1 6.178909e-12 -0.1544406    0.064 0.123 4.609466e-09 4.190424e-10\n",
       "MA0139.1 9.610053e-12 -0.1929606    0.607 0.691 7.169099e-09 5.974249e-10\n",
       "MA0740.1 1.212124e-11 -0.1424404    0.178 0.248 9.042443e-09 6.955725e-10\n",
       "MA0502.2 2.075256e-11  0.1943036    0.916 0.872 1.548141e-08 1.105815e-09\n",
       "MA0161.2 2.995669e-11  0.2212196    0.998 0.993 2.234769e-08 1.489846e-09\n",
       "MA1130.1 4.090674e-11 -0.2416664    0.005 0.020 3.051643e-08 1.840580e-09\n",
       "MA0162.4 4.194350e-11 -0.1424868    0.010 0.020 3.128985e-08 1.840580e-09\n",
       "MA1644.1 5.156797e-11  0.1658892    0.662 0.590 3.846971e-08 2.137206e-09\n",
       "MA0742.1 6.413254e-11 -0.1591836    0.131 0.199 4.784288e-08 2.518046e-09\n",
       "MA0060.3 1.348471e-10  0.1735255    0.651 0.587 1.005960e-07 4.814636e-09\n",
       "MA0593.1 1.355326e-10  0.1840209    0.887 0.833 1.011074e-07 4.814636e-09\n",
       "MA0477.2 1.965481e-10 -0.2164108    0.006 0.020 1.466249e-07 6.664766e-09\n",
       "MA1134.1 2.704817e-10 -0.2308074    0.007 0.020 2.017793e-07 8.773014e-09\n",
       "MA0072.1 2.961261e-10  0.2192463    0.988 0.979 2.209101e-07 9.204587e-09\n",
       "MA1116.1 3.534273e-10  0.1598859    0.929 0.886 2.636568e-07 1.054627e-08\n",
       "MA1137.1 3.878982e-10 -0.2284086    0.008 0.021 2.893721e-07 1.112970e-08\n",
       "MA0732.1 5.322822e-10 -0.1405313    0.022 0.038 3.970825e-07 1.470676e-08\n",
       "MA0666.1 8.867248e-10  0.1901239    0.964 0.952 6.614967e-07 2.100933e-08\n",
       "MA0708.1 8.867248e-10  0.1901239    0.964 0.952 6.614967e-07 2.100933e-08\n",
       "MA0709.1 8.867248e-10  0.1901239    0.964 0.952 6.614967e-07 2.100933e-08\n",
       "⋮        ⋮            ⋮             ⋮     ⋮     ⋮            ⋮           \n",
       "MA0814.2 0.8775416    -0.0012909307 0.648 0.646 1            0.9130349   \n",
       "MA1607.1 0.8907774     0.0120231042 0.597 0.606 1            0.9255152   \n",
       "MA0626.1 0.9003377     0.0121991967 0.126 0.139 1            0.9326034   \n",
       "MA0781.1 0.9005587     0.0024475142 0.354 0.364 1            0.9326034   \n",
       "MA0879.1 0.9015468    -0.0124056172 0.636 0.629 1            0.9326034   \n",
       "MA1569.1 0.9026001    -0.0002614462 0.476 0.467 1            0.9326034   \n",
       "MA0689.1 0.9083895     0.0035505919 0.164 0.195 1            0.9372871   \n",
       "MA0069.1 0.9212343    -0.0011142453 0.364 0.355 1            0.9492276   \n",
       "MA0819.1 0.9228512     0.0178486823 0.121 0.126 1            0.9495821   \n",
       "MA0815.1 0.9241424     0.0115731168 0.703 0.698 1            0.9496008   \n",
       "MA0839.1 0.9371164     0.0073612132 0.268 0.268 1            0.9616078   \n",
       "MA0024.3 0.9425644    -0.0013498868 0.526 0.526 1            0.9658696   \n",
       "MA0003.4 0.9444987    -0.0001510493 0.555 0.546 1            0.9665240   \n",
       "MA0811.1 0.9504210     0.0103099361 0.618 0.630 1            0.9712521   \n",
       "MA0037.3 0.9556658    -0.0023560906 0.138 0.160 1            0.9751980   \n",
       "MA1419.1 0.9569086     0.0135376607 0.616 0.612 1            0.9751980   \n",
       "MA0102.4 0.9582039    -0.0037102403 0.442 0.437 1            0.9751980   \n",
       "MA0767.1 0.9676795     0.0016362936 0.222 0.232 1            0.9834999   \n",
       "MA0028.2 0.9700230     0.0066278163 0.393 0.400 1            0.9845403   \n",
       "MA0749.1 0.9733881    -0.0009654235 0.406 0.402 1            0.9866135   \n",
       "MA1563.1 0.9793600     0.0039495215 0.529 0.533 1            0.9913196   \n",
       "MA1552.1 0.9824250    -0.0028840138 0.489 0.478 1            0.9927754   \n",
       "MA1471.1 0.9834598     0.0007837921 0.457 0.455 1            0.9927754   \n",
       "MA1483.1 0.9848484    -0.0015546475 0.419 0.428 1            0.9928336   \n",
       "MA1114.1 0.9874160     0.0044616199 0.228 0.243 1            0.9940787   \n",
       "MA0631.1 0.9895514    -0.0015982223 0.313 0.315 1            0.9948859   \n",
       "MA0847.2 0.9930888     0.0072745558 0.603 0.599 1            0.9959829   \n",
       "MA1576.1 0.9940583     0.0037482974 0.638 0.632 1            0.9959829   \n",
       "MA1554.1 0.9946478     0.0237875198 0.081 0.100 1            0.9959829   \n",
       "MA1573.1 0.9997445    -0.0040914399 0.624 0.615 1            0.9997445   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "\n",
    "# Perform differential testing \n",
    "cell_list <- c(\"Astrocytes\", \"ExNeuron\", \"InhNeuron\", \"Microglia\", \"Oligodendrocytes\", \"OPC\")\n",
    "\n",
    "cell_type <- cell_list[1] # CHANGE THIS FROM 1-6 TO PERFORM TEST ON EACH CELL TYPE\n",
    "\n",
    "test_name <- \"wilcox\"\n",
    "\n",
    "# EXAMPLE MARKERS for subset low high and celltype\n",
    "differential.activity <- FindMarkers(chromvar, assay=\"chromvar\", \n",
    "                                     ident.1 = 'cocaine.high',\n",
    "                                     ident.2 = 'cocaine.low',\n",
    "                                     group.by = 'condition',\n",
    "                                     subset.ident = cell_type,\n",
    "                                     test_use=test_name,\n",
    "                                     only.pos = FALSE,\n",
    "                                     logfc.threshold=0,\n",
    "                                     min.pct = 0,\n",
    "                                     mean.fxn = rowMeans,\n",
    "                                     fc.name = \"avg_diff\"\n",
    "                                    )\n",
    "\n",
    "# FDR adjust pvalues\n",
    "differential.activity$fdr = p.adjust(differential.activity$p_val, method='fdr')\n",
    "\n",
    "\n",
    "# Write diff data to outpur tsv\n",
    "diff_out <- sprintf(\"chromvar_diff_v2/%s_clusters_fdr.tsv\", cell_type)\n",
    "write.table(differential.activity, file=diff_out, sep=\"\\t\", row.names = TRUE, quote=FALSE)\n",
    "\n",
    "differential.activity # Show output data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "789e34b5-f1c1-4064-88fc-928a5a04e3d8",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R [conda env:sc_atac]",
   "language": "R",
   "name": "conda-env-sc_atac-r"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
